### PR TITLE
 Add Support for Volcano Jobs in Kubernetes Integration

### PR DESCRIPTION
**Pull request by zyn71**
# Add Support for Volcano Jobs in Kubernetes Integration

## Related Issue

Implements support for Volcano Jobs as requested in https://github.com/PrefectHQ/prefect/issues/17260

## Description

This PR adds support for [Volcano](https://volcano.sh) jobs in the Kubernetes worker integration. Volcano provides enhanced scheduling capabilities for high-performance computing workloads such as machine learning training, big data applications, and scientific computing.

### What is Volcano?

Volcano is a batch system built on Kubernetes providing mechanisms for high-performance workloads. It introduces a new custom resource (CRD) called `vcjob` that extends Kubernetes' capabilities with:

- Custom queue and priority scheduling
- GPU resource management
- Gang scheduling (minimum member requirements)
- Fine-grained lifecycle management
- Support for ML frameworks like PyTorch and TensorFlow

## Implementation Details

The implementation:

1. Adds conditional logic to detect Volcano job manifests by their `apiVersion: batch.volcano.sh/v1alpha1`
2. Updates methods to handle both standard Kubernetes jobs and Volcano jobs:
   - `_get_job`: Retrieves jobs using the appropriate API client
   - `_delete_job`: Deletes jobs using the appropriate API client
   - `_get_job_pod`: Locates pods using the `volcano.sh/job-name` label
   - `_find_pod_for_volcano_job`: Specialized method for finding Volcano job pods

3. Updates error handling to provide clear error messages for both job types

## Testing

The implementation has been tested with:
- Standard Kubernetes jobs (ensuring backwards compatibility)
- Volcano jobs with typical ML workloads
- Various pod discovery scenarios

## Notes for Reviewers

- The PR maintains backward compatibility with existing Kubernetes jobs
- Pod discovery for Volcano jobs relies on the `volcano.sh/job-name` label
- Error handling has been improved to provide clear guidance for both job types

## Screenshots (if applicable)

_Include any relevant screenshots showing the feature in action_

### CODE REVIEW

**Comment by zzstoatzz** [comment]
hi @zyn71 - looking at the linked issue #17260, I'm not sure this is the right direction. As a special case of kubernetes, the kubernetes worker shouldn't need to know anything about volcano. Rather as mentioned in that issue, we should be able to allow running volcano jobs by relaxing the validation logic that was preventing volcano jobs from running via a kubernetes work pool

I'll also note that the kubernetes worker is a widely used / important feature where changes must be well-tested.

let us know if you need any help!

**Comment by zyn71** [comment]
@zzstoatzz Thank you for your review and architectural guidance. As a university student with limited open source experience, I truly value this opportunity and your feedback. I completely understand and agree with the architectural principles you've raised - a more general approach without specific technology code is indeed better for maintenance.

I tried the approach you suggested of relaxing the validation logic and understand its value. During implementation, I encountered some technical challenges that I'd appreciate your further guidance on:

When relaxing validation to allow Volcano job manifests to pass through, I found several operational steps that still don't work properly:

1. The standard job management path uses `BatchV1Api`, but Volcano jobs require `CustomObjectsApi` for creation/deletion
2. Pod discovery for logs and monitoring fails because Volcano uses different label patterns (`volcano.sh/job-name` vs `job-name`)
3. Status monitoring needs to access different status fields in the job manifest

As a humble newcomer eager to learn, I'd like to ask for your advice on how to best handle these differences while maintaining a clean architecture. Some options I've considered:

- Would a more general "custom job handler" interface be appropriate?
- Perhaps a configuration-based approach where API paths and label selectors can be specified per job type?
- Or is there another pattern you'd recommend that I haven't considered?

I greatly appreciate your patient guidance and look forward to learning how to solve this problem in a more professional manner.

**Comment by zzstoatzz** [comment]
hi @zyn71 thanks for the context

I would suggest you check out [this](https://github.com/kevingrismore/volcano-worker) implementation of a volcano worker, as it should offer some insight. it may be true that the volcano worker has idiosyncrasies that warrant its own worker type, in which case I think we should create a `volcano` extra for `prefect_kubernetes` and have a parallel worker implementation. Before jumping there though, I would encourage you to review the linked impl and see if we can simplify your PR here

**Comment by zyn71** [comment]
Hi @zzstoatzz — thanks a lot for the suggestion and the helpful context!

I'll take a look at the implementation you linked and see how far we can go with that approach. Totally makes sense to avoid adding volcano-specific logic for now and try to make it work via configuration.

Appreciate the guidance — I’m excited to contribute and hopefully give something useful back to the community!

**Comment by zzstoatzz** [comment]
> I'll take a look at the implementation you linked and see how far we can go with that approach. Totally makes sense to avoid adding volcano-specific logic for now and try to make it work via configuration.
> 
> Appreciate the guidance — I’m excited to contribute and hopefully give something useful back to the community!

glad you're working on a contribution! 💙 

one thing I forgot to mention that would be helpful to look at is [the integration tests](https://github.com/PrefectHQ/prefect/tree/prefect-kubernetes-0.6.0/src/integrations/prefect-kubernetes/integration_tests/src/prefect_kubernetes_integration_tests) on the long-lived branch for the next minor version of `prefect-kubernetes`, an integration test in the style of the existing might be easier than the mocked unit tests on `main`

**Comment by zyn71** [comment]
Hi @zzstoatzz — thanks so much for your continued guidance!

Before I move forward, I’d love to quickly check if this direction aligns with the PR’s intent:

I plan to move all Volcano-specific logic into a new VolcanoWorker (prefect_kubernetes/volcano_worker.py),
so the existing KubernetesWorker can stay focused on batch/v1.

Two things I’m still unsure about:

Worker registration – would you prefer users start it via:
prefect worker start --pool my-pool --type volcano (register "volcano" as a worker type)
or use the full import path like
--type "prefect_kubernetes.volcano_worker.VolcanoWorker"?

apiVersion branching – I’m thinking the outer caller chooses the worker class,
and each worker only supports one apiVersion internally.
Is that what you had in mind, or would a small if/else wrapper at a higher level be better?

Also, just wanted to say — under your advice, we’ve already built and tested a version of this worker internally for our team,
and it’s working well so far. Really appreciate your support 🙏
We’d love to polish it further and contribute back in a way that fits the long-term direction of the project.

Thanks again, and let me know if I’m missing anything!

**Comment by zzstoatzz** [comment]
hi @zyn71 

> Worker registration – would you prefer users start it via:
prefect worker start --pool my-pool --type volcano (register "volcano" as a worker type)
or use the full import path like
--type "prefect_kubernetes.volcano_worker.VolcanoWorker"?

the former, the worker subclass should have a `type: str` classvar
```python
» uv run --extra kubernetes ipython

#[1]
from prefect_kubernetes.worker import KubernetesWorker

#[2]
KubernetesWorker.type
Out[2]: 'kubernetes'
```


> apiVersion branching – I’m thinking the outer caller chooses the worker class,
and each worker only supports one apiVersion internally.
Is that what you had in mind, or would a small if/else wrapper at a higher level be better?

what is the branching for (or maybe "where?" is a better question I might be misunderstanding, but if they're separate worker implementations (normal KubernetesWorker and VolcanoWorker) then when/why would a user need to select the worker class?
